pkg/kubectl/cmd/util/openapi/openapi.go

```diff
@@ -17,236 +17,108 @@ limitations under the License.
 )
 
 // Resources interface describe a resources provider, that can give you
 // resource based on group-version-kind.
 type Resources interface {
-	LookupResource(gvk schema.GroupVersionKind) Schema
-}
-
-// SchemaVisitor is an interface that you need to implement if you want
-// to "visit" an openapi schema. A dispatch on the Schema type will call
-// the appropriate function based on its actual type:
-// - Array is a list of one and only one given subtype
-// - Map is a map of string to one and only one given subtype
-// - Primitive can be string, integer, number and boolean.
-// - Kind is an object with specific fields mapping to specific types.
-// - Reference is a link to another definition.
-type SchemaVisitor interface {
-	VisitArray(*Array)
-	VisitMap(*Map)
-	VisitPrimitive(*Primitive)
-	VisitKind(*Kind)
-	VisitReference(Reference)
+	LookupResource(gvk schema.GroupVersionKind) proto.Schema
 }
 
-// Schema is the base definition of an openapi type.
-type Schema interface {
-	// Giving a visitor here will let you visit the actual type.
-	Accept(SchemaVisitor)
-
-	// Pretty print the name of the type.
-	GetName() string
-	// Describes how to access this field.
-	GetPath() *Path
-	// Describes the field.
-	GetDescription() string
-	// Returns type extensions.
-	GetExtensions() map[string]interface{}
-}
+// groupVersionKindExtensionKey is the key used to lookup the
+// GroupVersionKind value for an object definition from the
+// definition's "extensions" map.
+const groupVersionKindExtensionKey = "x-kubernetes-group-version-kind"
 
-// Path helps us keep track of type paths
-type Path struct {
-	parent *Path
-	key    string
+// Definitions is an implementation of `Resources`. It looks for
+// resources in an openapi Schema.
+type document struct {
+	// Maps gvk to model name
+	resources map[schema.GroupVersionKind]string
+	models    proto.Models
 }
 
-func NewPath(key string) Path {
-	return Path{key: key}
-}
+var _ Resources = &document{}
 
-func (p *Path) Get() []string {
-	if p == nil {
-		return []string{}
+func NewOpenAPIData(doc *openapi_v2.Document) (Resources, error) {
+	models, err := proto.NewOpenAPIData(doc)
+	if err != nil {
+		return nil, err
 	}
-	if p.key == "" {
-		return p.parent.Get()
-	}
-	return append(p.parent.Get(), p.key)
-}
-
-func (p *Path) Len() int {
-	return len(p.Get())
-}
-
-func (p *Path) String() string {
-	return strings.Join(p.Get(), "")
-}
-
-// ArrayPath appends an array index and creates a new path
-func (p *Path) ArrayPath(i int) Path {
-	return Path{
-		parent: p,
-		key:    fmt.Sprintf("[%d]", i),
-	}
-}
 
-// FieldPath appends a field name and creates a new path
-func (p *Path) FieldPath(field string) Path {
-	return Path{
-		parent: p,
-		key:    fmt.Sprintf(".%s", field),
+	resources := map[schema.GroupVersionKind]string{}
+	for _, modelName := range models.ListModels() {
+		model := models.LookupModel(modelName)
+		if model == nil {
+			panic("ListModels returns a model that can't be looked-up.")
```

Comment:
Why panic instead of error?

Reply:
Because it's really really not supposed to happen.